espwrap/adaptors/sendgrid_v3.py

```diff
@@ -248,15 +250,22 @@ def send(self):
             """
             send message and append response from this grp to list of returned responses for all grouped_recipients
             """
-            try:
-                logger.debug(message)
-                response = self.client.send(message)
-            except HTTPError as e:
-                logger.exception(_HTTP_EXC_MSG, self.subject, e.status_code, e.reason, e.body)
-            except Exception:
-                logger.exception('Unknown exception while sending a SendGridMassEmail.')
-            else:
-                responses.append(response)
+            attempt = 0
+            logger.debug(message)
+            while attempt < MAX_ATTEMPTS:
+                try:
+                    response = self.client.send(message)
+                    responses.append(response)
+                except TooManyRequestsError:
+                    attempt += 1
+                    logger.exception('[{}] Too many requests, trying to resend the message...'.format(attempt))
+                    time.sleep(1)
```

Comment:
The RateLimit error returns the time it requires to wait before being able to send emails again. To be on the safe side we should at the minimum sleep for that duration + buffer

What do you all think? or does everything think that 1 is sufficient because the limit is per second anyway?

In addition, rate limit lets us know how many were sent and how many can be sent. Should we use that data to make sure we don't "over" send?

Reply:
@shyaniv7 

It seems that not all endpoints return rate limit headers to lets us know how many were sent and how many can be sent
[This guy also did not receive these headers for send mail endpoint](https://stackoverflow.com/questions/43430416/sendgrid-send-email-rate-limit)

In case if rate limit headers are shown we can process next [headers](https://sendgrid.api-docs.io/v3.0/how-to-use-the-sendgrid-v3-api/api-rate-limits#rate-limit-response-header):
```
X-RateLimit-Limit: 500  # limit for some endpoint
X-RateLimit-Remaining: 499  # remaining requests for the endpoint
X-RateLimit-Reset: 1392815263  # looks like timestamp 
```
It is unobvious how we can use this information with the send-mail endpoint. Remaining rate limit for send-mail endpoint updates every second, for me it is overhead to try to handle it in more complex way

[Also here is discussion of implementation X-RateLimit headers and proposition to show Reset in delta-seconds instead of timestamp](https://github.com/sendgrid/docs/issues/5588) It makes sense in my opinion